extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java

```diff
@@ -1061,4 +1062,54 @@ public Locale getLanguageInfo() {
     public String getSupportInfo() {
         return "";
     }
+
+    @Nonnull
+    @Override
+    public List<StreamSegment> getStreamSegments() {
+        final ArrayList<StreamSegment> segments = new ArrayList<>();
+        try {
+            final JsonArray panels = initialData.getArray("engagementPanels");
+            JsonArray segmentArray = null;
+
+            // Search for correct JsonObject containing the data => panelIdentifier == engagement-panel-macro-markers
+            for (int i = 0; i < panels.size(); i++) {
+                if (panels.getObject(i).getObject("engagementPanelSectionListRenderer")
+                        .getString("panelIdentifier").equals("engagement-panel-macro-markers")) {
+                    segmentArray = panels.getObject(i).getObject("engagementPanelSectionListRenderer")
+                            .getObject("content").getObject("macroMarkersListRenderer").getArray("contents");
+                    break;
+                }
+            }
+
+            if (segmentArray != null) {
+                final long duration = getLength();
+
+                for (final Object object : segmentArray) {
+                    final JsonObject segmentJson = ((JsonObject) object).getObject("macroMarkersListItemRenderer");
+
+                    final int startTimeSeconds = segmentJson.getObject("onTap").getObject("watchEndpoint")
+                            .getInt("startTimeSeconds", -1);
+
+                    if (startTimeSeconds > duration) {
+                        // Rarely there are streams which show a greater length in the search results
+                        // than it's parsed on the details page, so ignore invalid segments set by uploader.
+                        break;
+                    }
+
+                    final String title = getTextFromObject(segmentJson.getObject("title"));
+                    final JsonArray previewsArray = segmentJson.getObject("thumbnail").getArray("thumbnails");
+
+                    final StreamSegment segment = new StreamSegment(title, startTimeSeconds);
+                    if (!previewsArray.isEmpty()) {
+                        // Assume that the thumbnail with the highest resolution is at the last position
+                        segment.setPreviewUrl(previewsArray.getObject(previewsArray.size() - 1).getString("url"));
+                    }
+                    segment.setUrl(getUrl() + "?t=" + startTimeSeconds);
+                    segments.add(segment);
+                }
+            }
+        } catch (Exception ignored) {
```

Comment:
why do you ignore all exceptions?

Reply:
The reason I decided to ignore them is that `getLength` method which could throw a `ParsingException` and because the segments are optional data, not like the title or description.

But thinking about it I should've worked with `.has()`  and return the empty list when it's "obviouly" not present by non-existing JSON-key, otherwise doing the actual parsing with its exceptions throwing if there are errors retrieving the objects etc.

I'd reset the three commits and do it more clearer tomorrow, okay? ;)